### PR TITLE
Build Tooling: Split JavaScript tasks to individual jobs

### LINKED ISSUE #10625: Make it easier for a contributor to determine what failed their build
Currently, it can be a bit difficult to determine why a build failed:

image

image

A couple of enhancements we could make:

1. More clearly summarize build failures at the end of each job.
2. In the Travis environment variables, document which tests are running so there's some simple reference of what might've failed.

From https://wordpress.slack.com/archives/C02RQC26G/p1539620166000100

### PR DESCRIPTION
**Pull request by aduth**
Related: #15159
Closes #10625

This pull request seeks to update the Travis configuration to run each individual script of the current "JS unit tests" job as its own separate job. As part of this, and as noted by @gziolo at https://github.com/WordPress/gutenberg/pull/14289#discussion_r265569094 , there is an added benefit here in that none of these scripts require `npm run build`, despite the fact that we run the build task. This has been removed here, and should help avoid any one of these containers from taking a particularly long time to run.

The proposed benefit is:

- Easier to determine what caused a build to fail
- Reduce the total time to run all of these tests, as each occurs in their own container
   - Note: There is some added overhead here in that each container must install its own dependencies.

**Testing instructions:**

Verify the build passes.

### CODE REVIEW

**Comment by danielbachhuber** [comment]
Sweet! Looks like there's still a failure though.

**Comment by aduth** [comment]
> Sweet! Looks like there's still a failure though.

Yeah, it seems the build may be required for the unit tests specifically after all. Or, at least, there's one package which has its own build script which needs to be run to generate an output relied upon by the test:

https://github.com/WordPress/gutenberg/blob/87d7fe4be406aedb3aefa7d961ac3781ab600654/packages/block-serialization-spec-parser/package.json#L29-L33

It may be enough to just do `npm lerna build` as a minimum.

This is accounted for normally in the pre-steps of `npm run build`:

https://github.com/WordPress/gutenberg/blob/87d7fe4be406aedb3aefa7d961ac3781ab600654/package.json#L158-L160

**Comment by gziolo** [comment]
> Yeah, it seems the build may be required for the unit tests specifically after all. Or, at least, there's one package which has its own build script which needs to be run to generate an output relied upon by the test:

Initially, it was integrated as `postinstall` step because it used to run builds for all packages. We can try to mirror it by adding `postinstall` script to the package itself and see if it solves the issue.

File: .travis.yml [hunk 1]
```diff
@@ -27,17 +27,36 @@ before_install:
...
   include:
-    - name: JS unit tests
-      env: WP_VERSION=latest
+    - name: Lint
       before_install:
         - nvm install --latest-npm
       install:
         - npm ci
       script:
-        - npm run build
```
Comment (gziolo) [inline .travis.yml]: Yes, ideally we don't need `npm run build` at all.

**Review by gziolo — COMMENTED** [review]
I think the initial idea was to have all those tasks in one job to minimize the impact of `npm install` command and all other bootstrap tasks involved.

**Comment by aduth** [comment]
> I think the initial idea was to have all those tasks in one job to minimize the impact of `npm install` command and all other bootstrap tasks involved.

Related: #14353

We removed the `postinstall` step because it was a slow and often unnecessary hurdle to installing dependencies. I think running `postinstall` for only custom `build` scripts of plugins wouldn't be quite so problematic, but it seems a bit puzzling that we'd perform some (but not all) of the build.

I guess it's by virtue of the Jest configuration [remapping for built files](https://github.com/WordPress/gutenberg/blob/f6f7375d8dfa9ef52ef784e5e486c71656421770/test/unit/jest.config.js#L13) (combined with `babel-jest`) that we're able to avoid `npm run build` at all. But this obviously doesn't cover build tasks _not_ accounted for by Babel. 

So it seems we could either:

1. "Cherry-pick" what we decide to build, i.e. as part of the Travis build, at least make sure custom packages are build (`npx lerna run build`)
2. Avoid possibility for overlooked build tasks by running the whole build (`npm run build`). This would obviously be slower (and partially unnecessary).

**Comment by aduth** [comment]
> 1. "Cherry-pick" what we decide to build, i.e. as part of the Travis build, at least make sure custom packages are build (`npx lerna run build`)

This is what I chose with 00358d8

**Comment by aduth** [comment]
While performance is not necessarily the primary objective here, some rough metrics look positive:

- Variable: https://travis-ci.com/WordPress/gutenberg/builds/111113034
- Control: https://travis-ci.com/WordPress/gutenberg/builds/110930966

Before:

-  JS unit tests: 6 min 40 sec

After:

- Lint: 2 min 3 sec
- Build artifacts: 1 min 38 sec
- License compatibility: 1 min 34 sec
- JavaScript unit tests: 5 min

While it uses 3 additional containers, it also saves roughly 1m40s as far as overall build time for these tasks (by longest runtime).

**Review by nerrad — APPROVED** [review]
These changes are looking good and especially like the splits which makes for easier "at a glance" feedback on any fails.

**Review by gziolo — APPROVED** [review]
This should work as suggested. I'm still not 100% confident that avoiding `lerna run build` in Gutenberg's `postinstall` is a good thing overall as it still may cause issues for developers who run `npm run test-unit` locally. We can tackle it separately as it's not the main goal of this PR.

Let's see how this new setup works in action.

**Comment by gziolo** [comment]
I have just realized that `@wordpress/block-serialization-spec-parser` is not used at all in the application. It exists only as a reference package which is used for comparison with other parsers. It is the only package which requires the build step which raises the question if we shouldn't rethink this setup and simplify it. Maybe this build step should be embedded in the test itself to ensure it always operates on the fresh versions of parser files generated.

**Pull request merged**